### ISSUE TITLE
[stable/prometheus-operator] PVC not mounted after prometheus restart

### ISSUE THREAD

**Issue by mikemowgli** [helm/charts#19762]
**Describe the bug**
After the prometheus pod is restarted, the PVC is not mounted, resulting in the queries.active file being unavailable

**Version of Helm and Kubernetes**:
Helm 3.0.0
Kubernetes 1.11 (OpenShift)

**Which chart**:
stable/prometheus-operator 8.3.3

**What happened**:

The chart was deployed successfully, prometheus was working properly. But if Kubernetes has to restart the pod, this error is yielded:
```
caller=query_logger.go:85 component=activeQueryTracker msg="Error opening query log file" file=/prometheus/queries.active err="open /prometheus/queries.active: permission denied"
--
  | panic: Unable to create mmap-ed active query log
```
After debugging the pod (preventing it to fail instantly by changing the command and args so I have a shell running), I noticed the PVC was not mounted, and as a result, the `/prometheus` direcotry belongs to root and is empty.
The events do not show that a volume cannot be attached. But they don't show either that a volume was attached.

**What you expected to happen**:

PVC is mounted after restart

**How to reproduce it**:

deploy the chart with persistent volumes for prometheus data, restart the pod

**Anything else we need to know**:
The only workaround I found was to manually delete the PVC, then run helm upgrade.

Here's an extract of my values file:
```
prometheus:
  enabled: true
  prometheusSpec:
    nodeSelector:
      node-role.kubernetes.io/infra: "true"
    replicas: 2
    walCompression: true
    retentionSize: 5.7GiB
    secrets:
    - etcd-client-cert
    securityContext:
      fsGroup: 1000120000
      runAsUser: 1000120000
    storageSpec:
      volumeClaimTemplate:
        spec:
          resources:
            requests:
              storage: 6Gi
```

I must admit I am not sure how the chart would be responsible for this, but as I see nothing wrong with other PVCs nor with my Storage Class, I may get some help here...

**Comment by mikemowgli** [comment]
The controller logs were:
```
I1224 13:27:34.846128       1 graph_builder.go:553] add [v1/Pod, namespace: monitoring, name: prometheus-prom-h3-prometheus-operato-prometheus-0, uid: 09bfdfbc-2651-11ea-9901-fa163ef6eac6] to the attemptToDelete, because it's waiting for its dependents to be deleted
I1224 13:27:34.846191       1 garbagecollector.go:408] processing item [v1/Pod, namespace: monitoring, name: prometheus-prom-h3-prometheus-operato-prometheus-0, uid: 09bfdfbc-2651-11ea-9901-fa163ef6eac6]
I1224 13:27:34.853215       1 garbagecollector.go:530] remove DeleteDependents finalizer for item [v1/Pod, namespace: monitoring, name: prometheus-prom-h3-prometheus-operato-prometheus-0, uid: 09bfdfbc-2651-11ea-9901-fa163ef6eac6]
I1224 13:27:38.866733       1 reconciler.go:229] attacherDetacher.DetachVolume started for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6" (UniqueName: "kubernetes.io/cinder/1f2e8d36-394b-4f60-8691-fd2dd86e0c01") on node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:38.869890       1 operation_generator.go:1235] Verified volume is safe to detach for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6" (UniqueName: "kubernetes.io/cinder/1f2e8d36-394b-4f60-8691-fd2dd86e0c01") on node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:39.812376       1 event.go:221] Event(v1.ObjectReference{Kind:"StatefulSet", Namespace:"monitoring", Name:"prometheus-prom-h3-prometheus-operato-prometheus", UID:"b381b4a8-12c3-11ea-8314-fa163ef6eac6", APIVersion:"apps/v1", ResourceVersion:"54001741", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' create Pod prometheus-prom-h3-prometheus-operato-prometheus-0 in StatefulSet prometheus-prom-h3-prometheus-operato-prometheus successful
I1224 13:27:40.041543       1 openstack_volumes.go:398] Successfully detached volume: 1f2e8d36-394b-4f60-8691-fd2dd86e0c01 from compute: b2b0214b-c35d-4132-80a6-3f72629fd266
I1224 13:27:42.645125       1 attacher.go:387] detached volume "1f2e8d36-394b-4f60-8691-fd2dd86e0c01" from node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:42.645170       1 operation_generator.go:419] DetachVolume.Detach succeeded for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6" (UniqueName: "kubernetes.io/cinder/1f2e8d36-394b-4f60-8691-fd2dd86e0c01") on node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:42.676204       1 reconciler.go:287] attacherDetacher.AttachVolume started for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6" (UniqueName: "kubernetes.io/cinder/1f2e8d36-394b-4f60-8691-fd2dd86e0c01") from node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:43.616943       1 openstack_volumes.go:362] Successfully attached 1f2e8d36-394b-4f60-8691-fd2dd86e0c01 volume to b2b0214b-c35d-4132-80a6-3f72629fd266 compute
I1224 13:27:47.613150       1 attacher.go:155] Attach operation successful: volume "1f2e8d36-394b-4f60-8691-fd2dd86e0c01" attached to instance "b2b0214b-c35d-4132-80a6-3f72629fd266".
I1224 13:27:47.724198       1 operation_generator.go:335] AttachVolume.Attach succeeded for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6" (UniqueName: "kubernetes.io/cinder/1f2e8d36-394b-4f60-8691-fd2dd86e0c01") from node "infra-node01-tools.openshift.dev.obfuscated.be"
I1224 13:27:47.724348       1 event.go:221] Event(v1.ObjectReference{Kind:"Pod", Namespace:"monitoring", Name:"prometheus-prom-h3-prometheus-operato-prometheus-0", UID:"2843bc3c-2651-11ea-9901-fa163ef6eac6", APIVersion:"v1", ResourceVersion:"54001761", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "pvc-51eac28e-1b28-11ea-9901-fa163ef6eac6"
```

As we can see, from the Kubernetes' perspective, everything is ok. What else can I check?

**Comment by mikemowgli** [comment]
After further investigation, we noticed this is not linked to the prometheus-operator, but because the PVC is full. Somehow we hit a bug that prevents mounting a completely full volume. As this point, we are not sure whether the bug comes from Kubernetes, OpenShift, or Docker. It doesn't seem related to our underlying OpenStack though, as the volume could be mounted to the VM, but not to the container.

Nevertheless, getting a full volume was a surprise, since in the Helm values, we used a combination of `prometheus.prometheusSpec.retentionSize` and `prometheus.prometheusSpec.retention`. But, as documented [in the Prometheus official doc](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects), the retention size flag is experimental, and does not account for the WAL size. In our case, the WAL size was over 2GB.

I'm closing this ticket, since the problem does not come from the chart, nevertheless, as mentioned in Prometheus doc, the retention size unit format is KB, MB, GB, and not `GiB` as in the README.

I'll open a new ticket for that.
